File: #basic-algorithms/sorting/Mergesort.java

public class Mergesort
{
	static void mergesort(int arr[], int left, int right)
	{
		if (left < right)
		{
			int middle = (left + right) / 2;
			mergesort(arr, left, middle);
			mergesort(arr, middle+1, right);
			merge(arr, left, middle, right);
		}
	}
	
	static void merge(int arr[], int left, int middle, int right)
	{
		int n1 = middle - left + 1;
		int n2 = right - middle;
		int L[] = new int[n1];
		int R[] = new int[n2];
		for (int i = 0; i < n1; i++)
		{
			L[i] = arr[left+i]; // notice. it is left+i rather than i
		}
		for (int j = 0; j < n2; j++)
		{
			R[j] = arr[middle+1+j];
		}
		int k = left, i = 0, j = 0; // notice. k=left rather than k = 0
		while(i < n1 && j < n2)
		{
			if (L[i] < R[j])
			{
				arr[k++] = L[i++];
			}
			else
			{
				arr[k++] = R[j++];
			}
		}
		
		while (i < n1)
		{
			arr[k++] = L[i++];
		}
		while (j < n2)
		{
			arr[k++] = R[j++];
		}
	}
	
	public static void main(String[] args)
	{
		int arr[] = {3, 1, 2, 4, 7, 4, 0, 3, 5}; // 0, 1, 2, 3, 3,  4, 4, 5, 7
		mergesort(arr, 0, arr.length - 1);
		for (int i = 0; i < arr.length; i++)
		{
			System.out.print(arr[i] + " ");
		}
		System.out.println();
	}
}

File: #basic-algorithms/sorting/Mergesort.cpp
#include <iostream>

using namespace std;

void merge(int arr[], int low, int mid, int high);

void mergesort(int arr[], int low, int high)
{
	if (low < high)
	{
		int mid = low + (high - low) / 2; // NOTICE it is low+(high-low)/2 rather than (high-low)/2
		mergesort(arr, low, mid);
		mergesort(arr, mid+1, high);
		merge(arr, low, mid, high);
	}
}

void merge(int arr[], int low, int mid, int high)
{
	int n = mid - low + 1;
	int m = high - mid;
	int a[n], b[m];
	for (int i = 0; i < n; i++)
	{
		a[i] = arr[low + i];
	}
	for (int j = 0; j < m; j++)
	{
		b[j] = arr[mid + j + 1];
	}
	int i = 0, j = 0, k = low;
	while (i < n && j < m)
	{
		if (a[i] <= b[j])
		{
			arr[k++] = a[i++];
		}
		else
		{
			arr[k++] = b[j++];
		}
	}
	while (i < n)
	{
		arr[k++] = a[i++];
	}
	while (j < m)
	{
		arr[k++] = b[j++];
	}
	
	for (int i = low; i <= high; i++)
	{
		cout << arr[i] << " ";
	}
	cout << endl;
}

int main()
{
	int n = 0;
	cout << "Please input size of array." << endl;
	cin >> n;
	cout << "Please input elements of array." << endl;
	int arr[n];
	for (int i = 0; i < n; i++)
	{
		cin >> arr[i];
	}
	mergesort(arr, 0, n - 1); // NOTICE. it is n-1 rather than n.
	for (int i = 0; i < n; i++)
	{
		cout << arr[i] << " ";
	}
	cout << endl;
	return 0;
}